### ISSUE TITLE
A question about ‘Loss is NaN’

### ISSUE THREAD

**Issue by yuejian607** [baronet2/SCISOR#3]
I'm very sorry to bother you again with another question.
I followed the steps on your Git repository for training, but encountered the error "Loss is NaN! Exiting!". 
I only modified a few places—for example, in the preprocess_uniref90_shard.pyscript, I changed processed = bp.process(batch, subbatch_size=512)(originally 64). I’m wondering if you’ve encountered similar issues before and how I can resolve this?

Attached is my uniref90.yamlconfiguration file:

data:
    data: uniref90
    N: 33
    train_path: /home/ubuntu/SCISOR/data/uniref90_processed
    max_segment: "013"
    n_val_shards: 2

model:
    seed: 0
    model: ShorteningSCUD
    forward_kwargs:
        pad_token_id: 1
        choices: [4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
        n_gen_samples: 10
        window_size: 2048
        alpha_rate: 1.1
        structured: true
        rao_blackwellize: true
    t_max: 0.9
    restart: false

architecture:
    nn_params:
        d_embedding: 128
        depth: 2
        slim: true
        hf_model_name: esm2_t12_35M_UR50D
        pretrained: True
    x0_model_class: FAESM

train:
    batch_size: 64
    n_epoch: 50
    lr: 0.0001
    lr_warmup: 1000
    grad_clip_val: 1
    weight_decay: 0
    accumulate: 8
    val_check_interval: 100000
    limit_val_batches: 5000   
    n_val_epochs: 5

sampling:
    gen_trans_step: 0

num_workers: 16

**Comment by baronet2** [comment]
Hi again!

Generally we found training SCISOR to be quite stable - how long into training did you encounter this issue? And are you also training on protein sequence data?

My advice would probably be the usual: maybe decrease the learning rate, add gradient clipping, etc. Perhaps also have a look at the numerical stability in some places (e.g. where there are logarithms or division) and increase the `eps`.

If the issue only happens after many batches of normal training (e.g. multiple hours of training), you can also try restarting the run from a checkpoint but using a different random seed.

Hope this helps!

**Comment by yuejian607** [comment]
Epoch 0:   1%|▍                           | 27078/1979694 [2:23:44<172:45:28,  3.14it/s, v_num=haof]
Loss is NaN! Exiting!
I  encountered the "Loss is NaN!" issue after reaching the above training progress. I am also training protein sequences and will try your suggestions. Thank you very much for your reply.

That said, I do have another question regarding your uniref90.yaml configuration: 
the val_check_intervalis :1000. 
Does this mean that for the uniref90 dataset, a test is performed every 1000 PyTorch files trained? Given that the training set contains an extremely large number of PyTorch files, would this result in hundreds of tests being conducted? Is my understanding correct?

**Comment by baronet2** [comment]
> Epoch 0: 1%|▍ | 27078/1979694 [2:23:44<172:45:28, 3.14it/s, v_num=haof]

Hmm yeah, that's annoying. Did the loss seem to be going down stably in the first 2 hours? Hope the suggestions I gave can fix this.

> Does this mean that for the uniref90 dataset, a test is performed every 1000 PyTorch files trained?

This is a pytorch lightning parameter. You can see the documentation [here](https://lightning.ai/docs/pytorch/stable/common/trainer.html#val-check-interval). In our experiments, we used this flag to run a validation epoch on some fixed validation data every so often, allowing us to monitor the model's progress. I think in our big training runs we aimed to have a validation cycle every few hours. You can adjust this and other lightning flags as you see fit.
